### PR TITLE
Record (pending): SP1024 Linear-Attention Hybrid + Late QAT + Hessian GPTQ

### PR DESCRIPTION
**Pull request by AVINASH0052**
## Summary

12-layer / 512-dim hybrid language model on SP1024 tokens.  Alternates
grouped-query softmax attention with a chunked gated linear-time mixer on
layers {1, 3, 5, 7, 9}.  Trained with Muon + AdamW for ~10 minutes on
8×H100, exported with Hessian-aware per-row GPTQ (int6 matrices, int7
embeddings) and Brotli-11 compression.  Sliding-window eval at stride 64.

## Architecture

- 12 transformer-style blocks at width 512; GQA 8q/2kv, head dim 64.
- Half-RoPE (first 32 of 64 head dims), sliding window 512 on the first
  8 layers and full attention on the top 4.
- Chunked gated linear-time mixer on odd layers: full softmax within
  64-token chunks plus a per-head low-rank state  \in \mathbb{R}^{d_k \times d_v}$ carried
  across chunks with a learned product-of-gates decay.
- SwiGLU MLP (expansion 5/2) with zero-initialised residual projections
  and per-dimension residual scales.
- Tied embedding head with tanh softcap 30.

## Training recipe (10-minute budget on 8×H100)

- Muon (Newton–Schulz 5-step) for 2-D weights; AdamW for embeddings and
  scalars, with separate LR groups for tied-embedding / head-scale / other.
- Trapezoidal LR: 60-step warmup → plateau → linear cool-down between
  steps 1900–2400.
- Grad-clip 1.0; Muon weight decay 0.05.
- EMA 0.9965 every 32 steps; SWA from step 2100 every 75 steps.
- Late QAT (fake-quant STE, int6) enabled from step 1950.

## Artifact pipeline

1. Merge SWA (fallback to EMA when no SWA checkpoint was collected).
2. Collect a diagonal-Hessian input-activation estimate on validation
   tokens for every `Linear` weight.
3. Hessian-aware per-row symmetric quantisation:
   - Matrices → int6 with role-specific σ-clip (3.0 MLP, 3.1 ATTN,
     3.2 MIXER).
   - Embedding → int7 with σ-clip 3.4.
   - Small 1-D tensors kept in fp16.
4. Bit-pack integer streams and Brotli-11 compress (bz2 fallback).
5. Round-trip decode, re-evaluate under sliding-window eval at stride 64
   with 1024-token context, and report
   `final_int8_zlib_roundtrip_exact val_loss val_bpb`.

## Reproduce

```bash
python3 data/cached_challenge_fineweb.py --variant sp1024
bash records/track_10min_16mb/2026-04-17_SP1024_LinearAttnHybrid/run_leaderboard_8xh100.sh
```

## Files

- `records/track_10min_16mb/2026-04-17_SP1024_LinearAttnHybrid/train_gpt.py` — single-file trainer (1230 lines, < 1500-line cap).
- `run_leaderboard_8xh100.sh` — production launcher.
- `run_smoke_1gpu.sh` — single-GPU sanity run.
- `submission.json` — metadata; will be populated with `val_bpb` + `bytes_total` after the 3-seed leaderboard run lands.

The submission is marked **pending compute** — score and `train.log`
will be added as follow-up commits once the 8×H100 run completes.

### CODE REVIEW

**Comment by AVINASH0052** [comment]
Closing: pivoting to a unique approach
